lumigator/python/mzai/jobs/inference/inference.py

```diff
@@ -0,0 +1,161 @@
+"""python job to run batch inference"""
+
+import argparse
+import functools
+import json
+import os
+import time
+from collections.abc import Iterable
+from pathlib import Path
+
+import s3fs
+from box import Box
+from datasets import load_from_disk
+from loguru import logger
+from model_clients import (
+    BaseModelClient,
+    MistralModelClient,
+    OpenAIModelClient,
+)
+
+# SummarizationPipelineModelClient,
+# HuggingFaceModelClient,
+from tqdm import tqdm
+
+
+def timer(func):
+    """Decorator which times the execution of the wrapped func.
+    Execution time is logged and also returned together with func's returned value
+    (output will be a tuple).
+    """
+
+    @functools.wraps(func)
+    def wrapper_timer(*args, **kwargs):
+        tic = time.perf_counter()
+        value = func(*args, **kwargs)
+        toc = time.perf_counter()
+        elapsed_time = toc - tic
+        logger.info(f"Elapsed time for {func.__name__}: {elapsed_time:0.4f} seconds")
+        return value, elapsed_time
+
+    return wrapper_timer
+
+
+@timer
+def predict(dataset_iterable: Iterable, model_client: BaseModelClient) -> list:
+    predictions = []
+
+    for sample_txt in dataset_iterable:
+        predictions.append(model_client.predict(sample_txt))
+
+    return predictions
+
+
+def save_outputs(config: Box, inference_results: dict) -> Path:
+    storage_path = config.evaluation.storage_path
+
+    def save_to_disk(local_path: Path):
+        logger.info(f"Storing into {local_path}...")
+        local_path.parent.mkdir(exist_ok=True, parents=True)
+        with local_path.open("w") as f:
+            json.dump(inference_results, f)
+
+    def save_to_s3(local_path: Path, storage_path: str):
+        s3 = s3fs.S3FileSystem()
+        if storage_path.endswith("/"):
+            storage_path = "s3://" + str(
+                Path(storage_path[5:]) / config.name / "inference_results.json"
+            )
+        logger.info(f"Storing into {storage_path}...")
+        s3.put_file(local_path, storage_path)
+
+    # generate local temp file ANYWAY
+    # (we don't want to lose all eval data if there is an issue wth s3)
+    local_path = Path(
+        Path.home() / ".lumigator" / "results" / config.name / "inference_results.json"
+    )
+
+    try:
+        save_to_disk(local_path)
+
+        # copy to s3 and return path
+        if storage_path is not None and storage_path.startswith("s3://"):
+            save_to_s3(local_path, storage_path)
+            return storage_path
+
+        return local_path
+
+    except Exception as e:
+        logger.error(e)
+
+
+def run_inference(config: Box) -> Path:
```

Comment:
Would using dictionaries or Pydantic model field access be sufficient here? I'm wary of adding levels of complexity if we don't need them.

Reply:
I agree on avoiding extra complexity, here I chose Box to reduce complexity at the expense of one extra dependency. The experience is similar to our previous pydantic configs: attribute-based access on nested values with a definable default, so we can try directly accessing a value and the worst that can happen is we get a None.

Ideally we'd have pydantic configs everywhere, but sharing them between jobs and lumigator definitely adds complexity and we are discussing the possibility of having them in lumigator only.

A simpler alternative would be to write some code to convert a nested dictionary using e.g. `types.SimpleNamespace`. Similar result with no dependencies, but without the default values for missing keys.

My 2 cents would be to continue using Box until we find a good solution for pydantic. WDYT?

Reply:
Per online discussion, let's take out box once we've aligned on a set of fields we'd like to access